src/wasi_unstable/mod.rs

```diff
@@ -526,99 +513,231 @@ pub fn path_symlink(old_path: &[u8], fd: Fd, new_path: &[u8]) -> Errno {
+#[inline]
+pub fn proc_raise(sig: Signal) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_proc_raise(sig) }
 }
 
-pub fn sock_recv(sock: Fd, ri_data: &[IoVec], ri_flags: RiFlags) -> (Errno, usize, RoFlags) {
+#[inline]
+pub fn sock_recv(sock: Fd, ri_data: &[IoVec], ri_flags: RiFlags) -> Result<(usize, RoFlags), Error> {
     let mut ro_datalen = MaybeUninit::<usize>::uninit();
     let mut ro_flags = MaybeUninit::<RoFlags>::uninit();
+
     unsafe {
-        (
-            __wasi_sock_recv(
-                sock,
-                ri_data.as_ptr(),
-                ri_data.len(),
-                ri_flags,
-                ro_datalen.as_mut_ptr(),
-                ro_flags.as_mut_ptr(),
-            ),
-            ro_datalen.assume_init(),
-            ro_flags.assume_init(),
-        )
+        let r = __wasi_sock_recv(
+            sock,
+            ri_data.as_ptr(),
+            ri_data.len(),
+            ri_flags,
+            ro_datalen.as_mut_ptr(),
+            ro_flags.as_mut_ptr(),
+        );
+        if let Some(code) = NonZeroU16::new(r) {
+            Err(code)
+        } else {
+            Ok((ro_datalen.assume_init(), ro_flags.assume_init()))
+        }
     }
 }
 
-pub fn sock_send(sock: Fd, si_data: &[CIoVec], si_flags: SiFlags) -> (Errno, usize) {
-    let mut so_datalen = MaybeUninit::<usize>::uninit();
-    unsafe {
-        (
-            __wasi_sock_send(
-                sock,
-                si_data.as_ptr(),
-                si_data.len(),
-                si_flags,
-                so_datalen.as_mut_ptr(),
-            ),
-            so_datalen.assume_init(),
-        )
-    }
+#[inline]
+pub fn sock_send(sock: Fd, si_data: &[CIoVec], si_flags: SiFlags) -> Result<usize, Error> {
+    unsafe_wrap!{ __wasi_sock_send(sock, si_data.as_ptr(), si_data.len(), si_flags) }
 }
 
-pub fn sock_shutdown(sock: Fd, how: SdFlags) -> Errno {
-    unsafe { __wasi_sock_shutdown(sock, how) }
+#[inline]
+pub fn sock_shutdown(sock: Fd, how: SdFlags) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_sock_shutdown(sock, how) }
 }
 
-pub fn sched_yield() -> Errno {
-    unsafe { __wasi_sched_yield() }
+#[inline]
+pub fn sched_yield() -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_sched_yield() }
 }
 
-pub fn fd_prestat_get(fd: Fd) -> (Errno, Prestat) {
-    let mut buf = MaybeUninit::<Prestat>::uninit();
-    unsafe {
-        (
-            __wasi_fd_prestat_get(fd, buf.as_mut_ptr()),
-            buf.assume_init(),
-        )
-    }
+#[inline]
+pub fn fd_prestat_get(fd: Fd) -> Result<Prestat, Error> {
+    unsafe_wrap!{ __wasi_fd_prestat_get(fd) }
 }
 
-pub fn fd_prestat_dir_name(fd: Fd, path: &mut [u8]) -> Errno {
-    unsafe { __wasi_fd_prestat_dir_name(fd, path.as_mut_ptr(), path.len()) }
+#[inline]
+pub fn fd_prestat_dir_name(fd: Fd, path: &mut [u8]) -> Result<(), Error> {
+    unsafe_wrap0!{ __wasi_fd_prestat_dir_name(fd, path.as_mut_ptr(), path.len()) }
+}
+
+#[cfg(feature = "alloc")]
+use alloc::{vec::Vec, string::String};
```

Comment:
Bringing in `Vec` and `String` makes me curious about the boundaries here. How do we decide what code belongs in this repo, versus in the Rust standard library?

The args and environ functions are low-level interfaces. The only reason I can think of why anyone would want to bypass the Rust standard library to call these would be if they really want to avoid any dynamic allocation, but in that case these `Vec`/`String` interfaces aren't appropriate either.

I'm also curious; if we do add these interfaces, will the Rust standard library will need to do its own dynamic allocation anyway, to copy the data into `OsString`s?

Reply:
Well, if you really don't want any allocations, there is still the raw interface. I guess we could reduce amount of allocations to just two by introducing a custom iterator, which will store the `argv` buffer.

An alternative would be to write something likes this:
```rust
fn get_args(f: impl FnMut(value: &[u8])) -> Result<(), Error> { .. }
```

In C we can use `alloca` to work around dynamic sizes, IIUC we could use [unsized rvalues](https://github.com/rust-lang/rust/issues/48055), but we still can't return anything unsized. Although with the callback approach unsized rvalus will allow us to make this function allocation-less. But I am not sure if we want make this crate nightly-only.

>will the Rust standard library will need to do its own dynamic allocation anyway, to copy the data into OsStrings

Yes, indeed with this function a total amount of allocations will be higher, so maybe idea with the custom iterator or callback is worth implementing.

Reply:
BTW does WASI provide any guarantees about strings encoding for args and env?

Reply:
I guess the question is, what's the motivating use case here? We do already have [a fine implementation of `args()`](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/wasi/args.rs#L30) in the standard library.

Currently, WASI does not define the encodings. I expect they'll be defined to be UTF-8 in the future, though we haven't had the time yet.

Reply:
How about API completeness? :) I've changed `get_args` (and `get_environ`) to a callback-based approach, so `std` implementation using it will have exactly the same number of allocations and in future we will be able to completely remove allocations from them.

Reply:
I have reworked env and args a bit, now API allows to pre-allocate buffer for them.